examples/dpo_demo_gemma3.ipynb

```diff
@@ -0,0 +1,849 @@
+    "    # greedy search\n",
+    "    \"greedy\": {\"temperature\": 1e-4, \"top_k\": 1, \"top_p\": 1.0},\n",
+    "    # some randomness\n",
+    "    \"standard\": {\"temperature\": 0.7, \"top_k\": 50, \"top_p\": 0.95},\n",
+    "    # liberal\n",
+    "    \"liberal\": {\"temperature\": 0.85, \"top_k\": 2000, \"top_p\": 1.0},\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "hnxxLtLAv66G"
+   },
+   "source": [
+    "# Load reference model and LoRA model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!huggingface-cli login"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model_id = \"google/gemma-3-1b-it\"\n",
+    "ignore_patterns = [\n",
+    "    \"*.pth\",  # Ignore PyTorch .pth weight files\n",
+    "]\n",
+    "print(f\"Downloading {model_id} from Hugging Face...\")\n",
+    "local_model_path = snapshot_download(\n",
+    "    repo_id=model_id, ignore_patterns=ignore_patterns\n",
+    ")\n",
+    "print(f\"Model successfully downloaded to: {local_model_path}\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def show_hbm_usage():\n",
+    "  \"\"\"Displays memory usage per device.\"\"\"\n",
+    "  fmt_size = functools.partial(humanize.naturalsize, binary=True)\n",
+    "\n",
+    "  print(\"\\n--- TPU HBM Usage ---\")\n",
+    "  for i, d in enumerate(jax.local_devices()):\n",
+    "    stats = d.memory_stats()\n",
+    "    used = stats.get(\"bytes_in_use\", 0)\n",
+    "    limit = stats.get(\"bytes_limit\", 0)\n",
+    "\n",
+    "    hbm_used = stats.get(\"device:0:HBM0:bytes_in_use\", used)\n",
+    "    hbm_limit = stats.get(\"device:0:HBM0:bytes_limit\", limit)\n",
+    "\n",
+    "    # Fallback if specific HBM stats not available\n",
+    "    if hbm_limit == 0:\n",
+    "      hbm_used = used\n",
+    "      hbm_limit = limit\n",
+    "\n",
+    "    percentage = (hbm_used / hbm_limit * 100) if hbm_limit > 0 else 0\n",
+    "\n",
+    "    print(\n",
+    "        f\"Device {i} ({d.device_kind}): Using {fmt_size(hbm_used)} /\"\n",
+    "        f\" {fmt_size(hbm_limit)} ({percentage:.2f}%)\"\n",
+    "    )\n",
+    "\n",
+    "  print(\"--- End HBM Usage ---\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(\"\\n--- HBM Usage BEFORE Model Load ---\")\n",
+    "show_hbm_usage()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "MODEL_CP_PATH = local_model_path\n",
+    "\n",
+    "model_config = (\n",
+    "    gemma3_model_lib.Gemma3Config.gemma3_1b()\n",
+    ")  # pick correponding config based on model version\n",
+    "MESH = [(1, 1), (\"fsdp\", \"tp\")]\n",
```

Comment:
MESH Is defined above in the config cell. In general, let's try to use stuff from the config cell, and remove vars which are not being used